### ISSUE TITLE
Resolve Pitest Issues - AvoidStarImportCheck (1)

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#7799]
Child issue of https://github.com/checkstyle/checkstyle/issues/7797 ,

This issue specifically focuses on the line https://github.com/checkstyle/checkstyle/blob/abf829f542023e089c741c55f73098cae52dcb12/.ci/pitest.sh#L74

**Comment by wilcoln** [comment]
I'm on it.

**Comment by wilcoln** [comment]
# 1. Pitest Report
- Link : https://wilcoln.github.io/checkstyle-reports/pit-reports/202003130421/index.html
- Surviving mutations:
     - on line 188: `removed conditional - replaced equality check with false → SURVIVED`  referenced by #7800 
     - on line 224: `removed conditional - replaced equality check with true → SURVIVED` we're focusing on it.

# 2. Hardcoded mutation

- checkstyle branch : https://github.com/wilcoln/checkstyle/tree/mutation-asic1

**Comment by wilcoln** [comment]
@rnveach, please review

**Comment by rnveach** [comment]
> The ast type is, indeed, STATIC_IMPORT
The ast type is IMPORT and allowClassImports is true
We're only interested in the 2nd case here, and in that case there's clearly no violation.

I agree that the 2nd one is the more important one. Regression was expanded to much properties, so I am good regression won't help us.

> Bottom line : I propose to keep the mutation.

Please convince me that the best choice is to remove the code, as there is no way this can produce any violation issues.

Why does, when the 2nd case's requirements are met, a violation is not produced? Do we have a duplicate check somewhere that checks for STATIC_IMPORT? What happens differently when ast is IMPORT versus STATIC_IMPORT that something else down the line won't produce a violation when the `else if` condition is met.

**Comment by wilcoln** [comment]
Let's assume the 2nd case's requirements are met for a some import statement.
1. The ast type is `IMPORT` which means the corresponding statement is a class import as opposed to a static member import.
2.  `allowClassImports = true` which means that, since our statement is a class import, we allow it.

That's why there is no violation.

**Comment by rnveach** [comment]
I agree with your number 1, but number 2 plays no role here. There are no further checks for `allowClassImports` inside the condition. That property only helped us get to the `else if`, nothing more.

**Comment by wilcoln** [comment]
> There are no further checks for `allowClassImports`
Yeah but since we're in case 2 of https://github.com/checkstyle/checkstyle/issues/7799#issuecomment-598535905, we know for sure that that property's value is `true`

> You can still have a class import with a start at this location

I agree with this completely, and what I am saying is that knowing `allowClassImports = true`, a class import with a star doesn't cause any violation.

At least that's what I thought from the property's description, but maybe I'm misunderstanding.

**Comment by wilcoln** [comment]
@rnveach I got proof!

I  just ran `java -jar checkstyle-8.29-all.jar -c config.xml Test.java` on the following source file:
```java
import java.util.*; // a class import with a star

class Foo{
  private int a;
  /* public comment */ // OK, comment is ignored
  public bar1() {} // violation
  public bar2() {} // violation
}
``
For each of the two following configs:
```xml
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
		<module name="AvoidStarImport"/>
```
And 
```xml
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
		<module name="AvoidStarImport">
  <property name="allowClassImports" value="true"/>
</module>
```

the results are respectively:

```
Starting audit...
Audit done.
```
And 

```
Starting audit...
[ERROR] /var/tmp/Test.java:2: Using the '.*' form of import should be avoided - java.util.*. [AvoidStarImport]
Audit done.
Checkstyle ends with 1 errors.
```

I believe this prove my point.

**Comment by rnveach** [comment]
> we know for sure that that property's value is true

Yes, but without it being used anymore it is not proof that it has any impact in the condition.

>  I got proof!

I don't see how this proves what is happening inside the code between 2 versions of checkstyle when you are only using 1 version in your example.  Surely regression would have had this case already.

-----

I suggest we walk through the code line by line inside the condition with a STATIC_IMPORT and IMPORT and see where the differences fall.

**Comment by wilcoln** [comment]
> I suggest we walk through the code line by line inside the condition with a STATIC_IMPORT and IMPORT and see where the differences fall.

Ok let's do that.
Let's consider the [mutation code](https://github.com/checkstyle/checkstyle/blob/4304bc2e894d9f8a4ae45ef47ca7287a8ded507f/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/AvoidStarImportCheck.java) (i.e there's no check if of the ast type inside the `else if` condition).
Let's assume the 2nd case's requirements are met for some import statement.
-  The ast type is `IMPORT`  and `allowClassImports` is `true`.
-  We get to the `else if` condition
    - If `allowStaticMembersImports` is `true` we do not enter the `else if` body => no violations
    - otherwise we do enter the `else if` body
      At this point, since ast type is `IMPORT` then at `line 255`  `startingDot` value is actually `SEMI (;)` and not `DOT (.)` 
Cf the tree below:
```
+--IMPORT (import)
    |
    +--DOT (.)
        |
        +--DOT (.)
            |
            +--IDENT (java)
            +--IDENT (io)
        +--IDENT (IOException)
    +--SEMI (;)
```
So we at `line 226` we call `logsStarredImportViolation(startingDot)` passing it a `SEMI(;)` instead of a `DOT(.)`.
Given that, `importText` local variable can't possibly end with `".*"` . That's the reason why we do not enter the `if` body at `line 238` Hence no violation is reported.

**Comment by wilcoln** [comment]
@rnveach , please your thoughts

**Comment by rnveach** [comment]
This makes more sense now.

> Given that, importText local variable can't possibly end with ".*". Hence we do not enter the if body at line 238

There was a small leap not saying how importText got its value but its directly from the AST brought in which is only a `;` in the bad case. I was able to still follow.

I agree now, there is no way to satisfy the code as it is now except to change it somehow, either remove the mutated code or rewrite the code.

**Comment by rnveach** [comment]
Fix was merged
